File: src/simulation/culture/thinking/meaning/GroupMemes.java
package simulation.culture.thinking.meaning;

import kotlin.Pair;
import shmp.random.RandomCollectionsKt;
import shmp.random.RandomProbabilitiesKt;
import simulation.Controller;
import simulation.culture.aspect.Aspect;
import simulation.space.resource.Resource;

import java.util.*;
import java.util.stream.Collectors;

public class GroupMemes extends MemePool {
    private Map<String, Meme> memesCombinations;

    public GroupMemes() {
        super();
        memesCombinations = new HashMap<>();
        String[] subjects = {"group", "time", "space", "life", "death", "sun", "luck", "misfortune"};
        addAll(Arrays.stream(subjects).map(MemeSubject::new).collect(Collectors.toList()));
        String[] predicates = {"die", "acquireAspect", "consume", "exist", "and"};
        addAll(Arrays.stream(predicates).map(MemePredicate::new).collect(Collectors.toList()));
        addAll(Controller.session.templateBase.wordBase.values().stream()
                .flatMap(Collection::stream)
                .map(Meme::copy).collect(Collectors.toList()));
    }

    public void addAll(GroupMemes groupMemes) {
        super.addAll(groupMemes);
        groupMemes.memesCombinations.values().forEach(this::addMemeCombination);
    }

    @Override
    public List<Meme> getMemes() {
        List<Meme> memeList = super.getMemes();
        memeList.addAll(memesCombinations.values());
        return memeList;
    }

    @Override
    public Meme getMeme(String name) {
        Meme memeS = super.getMeme(name);
        return memeS != null ? memeS : getMemeCombinationByName(name);
    }

    @Override
    public Meme getMemeCopy(String name) {
        Meme meme = super.getMemeCopy(name);
        if (meme != null) {
            return meme;
        }
        meme = getMemeCombinationByName(name);
        return meme == null ? null : meme.copy();
    }

    private Meme getMemeCombinationByName(String name) {
        return memesCombinations.get(name.toLowerCase());
    }

    public Meme getValuableMeme() {
        return chooseMeme(getMemes());
    }

    public Meme getMemeWithComplexityBias() {
        if (RandomProbabilitiesKt.testProbability(0.5, Controller.session.random)) {
            return getValuableMeme();
        }
        return chooseMeme(new ArrayList<>(memesCombinations.values()));
    }

    private Meme chooseMeme(List<Meme> memeList) {
        return RandomCollectionsKt.randomElement(
                memeList,
                m -> ((double) m.importance) * ((double) m.importance),
                Controller.session.random
        );
    }

    @Override
    public boolean isEmpty() {
        return super.isEmpty() && memesCombinations.isEmpty();
    }

    public void addAspectMemes(Aspect aspect) {
        addPairMemes(ConstructMemeKt.makeAspectMemes(aspect));
    }

    public void addResourceMemes(Resource resource) {
        addPairMemes(ConstructMemeKt.makeResourceMemes(resource));
    }

    public void addPairMemes(Pair<List<Meme>, List<Meme>> memes) {
        memes.getFirst().forEach(this::add);
        memes.getSecond().forEach(this::addMemeCombination);
    }

    public void addMemeCombination(Meme meme) {
        if (!memesCombinations.containsKey(meme.toString())) {
            memesCombinations.put(meme.toString(), meme.copy());
        }
    }

    @Override
    public boolean strengthenMeme(Meme meme) {
        return strengthenMeme(meme, 1);
    }

    public boolean strengthenMeme(Meme meme, int delta) {
        if (super.strengthenMeme(meme, delta)) {
            return true;
        }
        Meme existing = getMemeCombinationByName(meme.toString());
        if (existing != null) {
            existing.increaseImportance(delta);
            return true;
        }
        if (meme.isSimple()) {
            add(meme.copy());
            return true;
        }
        addMemeCombination(meme.copy());
        return true;
    }
}

File: src/simulation/culture/thinking/meaning/GroupMemes.kt
package simulation.culture.thinking.meaning

import shmp.random.randomElement
import shmp.random.testProbability
import simulation.Controller
import simulation.Controller.*
import simulation.culture.aspect.Aspect
import simulation.space.resource.Resource
import kotlin.math.pow


class GroupMemes : MemePool() {
    private val memesCombinationsMap = mutableMapOf<String, Meme>()

    private var popularMemes = listOf<Meme>()

    private fun updatePopular() {
        popularMemes = all.sortedByDescending { it.importance }.take(50)
    }

    init {
        addAll(
                listOf("group", "time", "space", "life", "death", "sun", "luck", "misfortune")
                        .map { MemeSubject(it) }
        )

        addAll(
                listOf("die", "acquireAspect", "consume", "exist", "and")
                        .map { MemePredicate(it) }
        )

        addAll(
                session.templateBase.wordBase.values
                        .flatten()
                        .map { it.copy() }
        )

        updatePopular()
    }

    override val isEmpty: Boolean
        get() = super.isEmpty && memesCombinationsMap.isEmpty()

    fun addAll(groupMemes: GroupMemes) {
        super.addAll(groupMemes)

        groupMemes.memesCombinationsMap.values
                .forEach { addMemeCombination(it) }
    }

    override val all get() = memesMap.values + memesCombinationsMap.values

    override fun getMeme(name: String) = super.getMeme(name)
            ?: getMemeCombinationByName(name)

    override fun getMemeCopy(name: String): Meme? {
        super.getMemeCopy(name)?.let { return it.copy() }

        return getMemeCombinationByName(name)?.copy()
    }

    private fun getMemeCombinationByName(name: String) =
            memesCombinationsMap[name.toLowerCase()]//TODO doublewut

    val valuableMeme: Meme
        get() {
            if (testProbability(0.9, session.random))
                return chooseMeme(popularMemes)

            updatePopular()

            return chooseMeme(all)
        }

    val memeWithComplexityBias: Meme
        get() =
            if (testProbability(0.5, session.random))
                valuableMeme
            else
                chooseMeme(memesCombinationsMap.values.toList())

    private fun chooseMeme(memeList: List<Meme>): Meme {
        return randomElement(
                memeList,
                { it.importance.toDouble().pow(2) },
                session.random
        )
    }

    fun addAspectMemes(aspect: Aspect) = addPairMemes(makeAspectMemes(aspect))

    fun addResourceMemes(resource: Resource) = addPairMemes(makeResourceMemes(resource))

    fun addPairMemes(memes: Pair<List<Meme>, List<Meme>>) {
        memes.first.forEach { add(it) }
        memes.second.forEach { addMemeCombination(it) }
    }

    fun addMemeCombination(meme: Meme) {
        if (!memesCombinationsMap.containsKey(meme.toString()))
            memesCombinationsMap[meme.toString()] = meme.copy()
    }

    override fun strengthenMeme(meme: Meme)
            = strengthenMeme(meme, 1)

    override fun strengthenMeme(meme: Meme, delta: Int): Boolean {
        if (super.strengthenMeme(meme, delta))
            return true

        getMemeCombinationByName(meme.toString())
                ?.let {
                    it.increaseImportance(delta)
                    return true
                }

        if (meme.isSimple) {
            add(meme.copy())
            return true
        }

        addMemeCombination(meme.copy())
        return true
    }
}
